rfc-106-docker-for-local-development.md

```diff
@@ -0,0 +1,37 @@
+# RFC 106: Use Docker for local development
+
+## Summary
+
+Adopt a Docker-based approach for local development instead of the Vagrant VM.
+
+## Problem
+
+We currently rely on the Vagrant development VM for the day to day dev cycle.
+
+- It has [many documented issues](https://docs.publishing.service.gov.uk/manual.html#development-vm)
+- Changes to puppet often result in a broken VM, because the infrastructure team don't use it.
+- We are adopting more cloud-based services into our stack, like S3 and Amazon's Elasticsearch. This means that our development environment doesn't look like production anymore, but we still need code in govuk-puppet.
+- It's hard to keep updated. Running the puppet command will often error and take a long time.
+- GOV.UK is moving to a containerised infrastructure
+
+## Possible approaches to local development
+
+- An approach based on **Docker and docker-compose**. Ben Thorner has a working prototype called [govuk-docker](https://github.com/benthorner/govuk-docker) and we have the [end-to-end tests](https://github.com/alphagov/publishing-e2e-tests) running in the same. Other GDS programmes like GOV.UK Pay use this approach.
+- An approach based on **GNU Guix** package manager, which Chris Baines uses via govuk-guix.
+- Because they are read-only, frontend applications on GOV.UK can be **run against production** using the startup.sh --live flag.
+- Developers on the GOV.UK PaaS work by developing against a **remote set of services** - we’d run many versions of GOV.UK in the cloud, which developers can use in development.
+- Applications can be run locally by **installing dependencies manually**. This is what some of GOV.UK’s Linux users do.
+
+
+Approach | Pro | Cons
+-- | -- | --
+Vagrant VM | Can look like production | Resource intensive, brittle
+Docker | Aligns with future hosting. Highly reproducible. | Slow because on a Mac it needs a VM. Development won’t match production exactly until we migrate Production onto a container-based platform.
+Guix | Huge amount of work put in | Technology unknown in GOV.UK
+Against production | Low setup, fast | Only works for frontend applications. Still needs local install
+Remote dependencies | Low setup. Can match Production exactly in terms of technologies used | Expensive, no offline development
+Manual install | Fast in development | Lots of setup
```

Comment:
I think we shouldn't be too quick to dismiss this option, particular for our more simple backend applications. This is generally the method I use to work and it's often the quickest and easier way for me to run tests.

```
$ brew install postgres redis
$ brew services start postgres
$ brew services start redis
$ bundle exec rake
```

This method certainly isn't appropriate for all our applications (particular ones with large numbers of dependencies) or for everyone. It can be quite daunting and sometimes causes problems with it not being well documented, but for a lot of our apps, I think it's often the simplest option.

Reply:
Another issue for manual install is we'd have to consider supporting multiple platforms - developers are starting use other platforms such as Ubuntu, Guix etc... but agree for running tests its been the easiest.

Reply:
I'm not sure whether we'd *need* to support multiple platforms.  There would be the blessed "this is how you run GOV.UK locally on your GDS-issued Mac" approach, with some documented changes for Linux users, but anything more exotic (Guix, Nix, etc) would be up to the individual to figure out.

Reply:
I've been using the local approach too, for a long time. The main drawback was getting apps to talk to each other locally - often the DB is the easiest thing to get working. Eventually I got to the point where I just wanted one dev environment that could do everything well.

Reply:
I think there's a lot to be said for a manual install. I've seen a few people struggle with the dev vm because they mistyped a command and ran in on their host rather than their vm, or were just confused about when they should be using the vm and when not..... A manual install removes this class of problem. The advantage of a VM though is (ideally) it's cheap and easy to throw away if you want to start fresh. That can be harder to do with a manual local install. Probably more so with things like postgres/redis/elasticsearch/rabbit than with the actual code and ruby versions to be fair.

Reply:
I also strongly agree with @barrucadu - we should support the default case of people developing on a GDS-issued Mac. That should be the golden path. The GOV.UK way if you like. If you want to veer from it then that's fine, but it's unofficial. Part of the expectation then, is you accept you'll need to be somewhat self sufficient and not expect the community to support your setup.

Reply:
I like developing locally too. It's super nice to work on frontend apps because they can run locally, get data from production and don't need any backing services.

One thing I was thinking about is whether we could run the app you're working on locally, but run the dependencies in Docker. Theoretically you'd only need rbenv on your machine, which limits the number of things that can go wrong. Do we think this could be a secondary supported way of running?

Reply:
That's what I've been doing for running search-api tests outside the VM: dockerised elasticsearch + rbenv.

Reply:
Yup agree - we wouldn't need to support _everything_, but as we're starting to see GDS issued Linux machines, we _might_ need to consider more than just MacOS as an "official" platform.

Reply:
Just to add that the GDS-issues Linux laptops are not "official" in the same sense that the Macs are - the Macs are still the default machine, and there is work ongoing to work out how various GDS internal services and software runs on these machines. So we should try but not kill ourselves attempting to support every flavour of Linux that may run on these machines.

Reply:
Good point, however the firebreak work that Will P did some time back with licensify provided a useful benchmark for docker performance on mac compared to linux. From what I remember compiling and running end to end tests took about 40 minutes on mac compared to about 4 . minutes on ubuntu linux, its was about an order of magnitude faster. @stephenharker has some context on this sine he uses linux.

Reply:
I think we should be careful about using Licensify as an example, since it's such an outlier on GOV.UK anyway. While it's true that Docker is slower on Mac than on Linux, where it's as fast as running stuff natively, the impact depends very much on [how we choose to configure it](https://github.com/benthorner/govuk-docker/blob/master/DECISIONS.md#docker-bindmount), which includes utilising recent performance improvements from the Docker developers.

As someone who now uses Docker for everything except editing, it might be reassuring to know it performs well enough to be a contender for the Dev VM. But it is true that running stuff natively on Mac will be faster.

Reply:
@benthorner I mentioned licensify more of an example of something where there was some 'prior art' that might be work taking a look at.